src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,552 @@
+open Core_kernel
+open Import
+open Snarky
+open Snark_params
+open Snark_params.Tick
+open Let_syntax
+open Currency
+open Fold_lib
+open Snark_bits
+
+let coinbase_tree_depth = Snark_params.pending_coinbase_depth
+
+(*Total number of stacks*)
+let coinbase_stacks = Int.pow 2 coinbase_tree_depth
+
+module Coinbase_data = struct
+  type t = Public_key.Compressed.t * Amount.Signed.t [@@deriving sexp]
+
+  let of_coinbase (cb : Coinbase.t) : t Or_error.t =
+    Option.value_map cb.fee_transfer
+      ~default:(Ok (cb.proposer, Amount.Signed.of_unsigned cb.amount))
+      ~f:(fun (_, fee) ->
+        match Currency.Amount.sub cb.amount (Amount.of_fee fee) with
+        | None -> Or_error.error_string "Coinbase underflow"
+        | Some amount -> Ok (cb.proposer, Amount.Signed.of_unsigned amount) )
+
+  type var = Public_key.Compressed.var * Amount.Signed.var
+
+  type value = t
+
+  let length_in_triples =
+    Public_key.Compressed.length_in_triples + Amount.Signed.length_in_triples
+
+  let typ : (var, value) Typ.t =
+    let spec =
+      let open Data_spec in
+      [Public_key.Compressed.typ; Amount.Signed.typ]
+    in
+    let of_hlist : 'a 'b. (unit, 'a -> 'b -> unit) H_list.t -> 'a * 'b =
+      let open H_list in
+      fun [public_key; amount] -> (public_key, amount)
+    in
+    let to_hlist (public_key, amount) = H_list.[public_key; amount] in
+    Typ.of_hlistable spec ~var_to_hlist:to_hlist ~var_of_hlist:of_hlist
+      ~value_to_hlist:to_hlist ~value_of_hlist:of_hlist
+
+  let var_of_t ((public_key, amount) : value) =
+    ( Public_key.Compressed.var_of_t public_key
+    , Amount.Signed.Checked.of_unsigned @@ Amount.var_of_t
+      @@ Amount.Signed.magnitude amount )
+
+  let var_to_triples (public_key, amount) =
+    let%map public_key = Public_key.Compressed.var_to_triples public_key in
+    let amount = Amount.Signed.Checked.to_triples amount in
+    public_key @ amount
+
+  let fold ((public_key, amount) : t) =
+    let open Fold in
+    Public_key.Compressed.fold public_key +> Amount.Signed.fold amount
+
+  let crypto_hash_prefix = Hash_prefix.coinbase
+
+  let crypto_hash t = Pedersen.hash_fold crypto_hash_prefix (fold t)
+
+  let empty = (Public_key.Compressed.empty, Amount.zero)
+
+  let digest t = Pedersen.State.digest (crypto_hash t)
+
+  let create public_key amount = (public_key, amount)
+
+  let gen =
+    let open Quickcheck.Let_syntax in
+    let%bind public_key = Public_key.Compressed.gen in
+    let%bind amount = Currency.Amount.gen in
+    return (create public_key amount)
+
+  module Checked = struct
+    let hash t =
+      var_to_triples t
+      >>= Pedersen.Checked.hash_triples ~init:crypto_hash_prefix
+
+    let digest t =
+      var_to_triples t
+      >>= Pedersen.Checked.digest_triples ~init:crypto_hash_prefix
+  end
+end
+
+module Stack_pos = struct
```

Comment:
maybe put a type signature on this to make sure you don't accidentally turn a too-large int into a stack_pos?